### ISSUE TITLE
Use ABL loss to Binary classification

### ISSUE THREAD

**Issue by Buling-Knight** [wangchi95/active-boundary-loss#8]
Thank you very much for your excellent work!!! In the process of using, I have a question。
when I try to use it to binary classification task, I modified this line of code, it looks work. I wonder if I should note other lines?
![1](https://user-images.githubusercontent.com/91008145/201809634-6ac9aea7-506b-40e8-b22c-7042c8612263.jpg)

**Comment by moonfoam** [comment]
It should not work well since `p` and `q` in the KL divergence should both be a distribution, which is guaranteed with the softmax operation here. (For input with shape [Batch, 1, H, W], you can also take the concatenation of the sigmoid output `t` and (1-`t`) to keep the sum of dim 1 is 1.)  If you have further questions, you can post the code of your loss and the shape of its input data.

**Comment by wangchi95** [comment]
It should not work well since `p` and `q` in the KL divergence should both be a distribution, which is guaranteed with the softmax operation here. (For input with shape [Batch, 1, H, W], you can also take the concatenation of the sigmoid output `t` and (1-`t`) to keep the sum of dim 1 is 1.) If you have further questions, you can post the code of loss and the shape of its input data.

**Comment by Buling-Knight** [comment]
@wangchi95 ,thanks for your reply and suggestions.
 I modified the code as the following Img1. And I got some error in the epoch 10(showed in img2). 
I noted that you tell me  "if pred_boundary.sum() < 1:   return None" int the code. 

My problem is that I nenver got this error in the same dataset when I set the decode output [b,1, 2, h, w] (now it's [b, 1, h, w]).
"if  pred_boundary.sum()< 1 ,skip the loss". can it work well?

![1](https://user-images.githubusercontent.com/91008145/201922616-b27f5052-6a72-448f-811a-77f51e47acd8.jpg)


![2](https://user-images.githubusercontent.com/91008145/201921784-930d02dc-250a-400c-a465-a877f6a9d2b0.jpg)
http
![3](https://user-images.githubusercontent.com/91008145/201922251-c8878dab-729d-433a-a3f5-73d8fef025dc.jpg)
s://user-images.githubusercontent.com/91008145/201921768-f3f5a268-6524-4a71-9ac0-a44fe77af533.jpg)

**Comment by wangchi95** [comment]
Oh, I got your point. Since the `pred` is logits (we don't know a certain max value), maybe you can concat `-pred` and `pred` to make it work.
In addition, you can ignore the ABL in the step with `None` output, which means there detect no obvious prediction boundary. And the boundary is detected here:
https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L65

And if you are sure there is an obvious boundary there, you can modify this threshold.

https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L38
https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L66

**Comment by Buling-Knight** [comment]
Thank you for your prompt reply，I'll try to modify it as your suggestions.

**Comment by JosephKKim** [comment]
Hi, I am also using the ABL for the binary segmentation, 
I made something like boundary_loss which pretty much like the one Buling-Knight has made.

I made two variables new_pred, and label_pred and each variable has the shape [1, 2, 592, 592].

While calculating the loss, function gt2boundary gives the error message like

> The size of tensor a (592) must match the size of tensor b (593) at non-singleton dimension 3

The shape gt_lr and gt_ud in the function is: gt_lr:  torch.Size([1, 1, 593, 592])
gt_ud:  torch.Size([1, 2, 591, 593]). 

I think I haven't understand the function correctly, and  I don't understand why the shape  looks like I've written above. 

Thanks for the nice work!
Joseph.

**Comment by wangchi95** [comment]
> Hi, I am also using the ABL for the binary segmentation, I made something like boundary_loss which pretty much like the one Buling-Knight has made.
> 
> I made two variables new_pred, and label_pred and each variable has the shape [1, 2, 592, 592].
> 
> While calculating the loss, function gt2boundary gives the error message like
> 
> > The size of tensor a (592) must match the size of tensor b (593) at non-singleton dimension 3
> 
> The shape gt_lr and gt_ud in the function is: gt_lr: torch.Size([1, 1, 593, 592]) gt_ud: torch.Size([1, 2, 591, 593]).
> 
> I think I haven't understand the function correctly, and I don't understand why the shape looks like I've written above.
> 
> Thanks for the nice work! Joseph.

The GT (maybe label_pred in your code) should has the shape [N, H, W] (not in one-hot format, should be [1,592,592] here). And the shape of `gt_lr` and `gt_ud` should both be [1,592,592] after padding.

The function `gt2boundary` is used to extract the boundary by comparing each pixel with its neighboring pixels to the right and below (1 if they have exactly the same label, 0 otherwise).

Please note that the notation `lr` and `ud` refer to left-right and up-down directions, respectively. Sorry I seem to be using the wrong variable names (the `gt_lr` and ` gt_ud` should be swapped), Although this mistake does not affect the final results, it may cause difficulties in understanding. I will correct this naming mistake now.

Best regards,
Chi Wang

**Comment by JosephKKim** [comment]
Thanks for the quick reply!
Oh I deleted the previous issue because I figured it out... Sorry for the deleting the previous issue I made. 
I solved it with making the GT, [1, 592, 592] with squeeze, which is just the [b, H, W]

I am also getting the return None  problem just like what Buling-Knight experienced. 
You told Buling-Knight to modify the threshold. I guess this error is due to the hard-to-find prediction boundary in the early stage of the training. 

In this case, what how should I modify those parameters to predict return None case?


I think eps is for detecting the PDB so eps should go down, and max_N should decrease because it is ratio of the # of boundary pixels out of the total pixels?
Am I correct???

**Comment by JosephKKim** [comment]
Plus, I found out that I am training image using patches, scale is like 50 x 50, in this case where should I change the code? 

your answer will be really help for the research.
Thanks in advance~

**Comment by wangchi95** [comment]
> Thanks for the quick reply! Oh I deleted the previous issue because I figured it out... Sorry for the deleting the previous issue I made. I solved it with making the GT, [1, 592, 592] with squeeze, which is just the [b, H, W]
> 
> I am also getting the return None problem just like what Buling-Knight experienced. You told Buling-Knight to modify the threshold. I guess this error is due to the hard-to-find prediction boundary in the early stage of the training.

Yes, you are right, there should better have different thresholds for different stages. So I use the max_N to adaptively adjust the thresholds. There are 2 cases of `None`: 1. there is no detected boundary. 2. all detected boundaries are discarded due to correct prediction (because it is on the GT boundary).
There are many reasons for producing case 1. In addition to the reasons you said, some datasets contain images without borders, so I did not forcibly extract the borders.

> In this case, what how should I modify those parameters to predict return None case?
> 
> I think eps is for detecting the PDB so eps should go down, and max_N should decrease because it is ratio of the # of boundary pixels out of the total pixels? Am I correct???

Yes, you can decrease eps gradually. However, I do not think you should decrease `max_N` during training, since the alignment between PDB and GTB should get better with training. If `max_N` is reduced, this loss function may not provide gradients since all high-confidence boundaries are predicted correctly. So I recommend fixing the value of `max_N` according to your dataset.

**Comment by wangchi95** [comment]
> Plus, I found out that I am training image using patches, scale is like 50 x 50, in this case where should I change the code?
> 
> your answer will be really help for the research. Thanks in advance~

You don't need to modify the code except to modify `max_N_ratio` according to the dataset. (Maybe slightly increase it.)

Hope this helps you:)

**Comment by JosephKKim** [comment]
Once again, thanks for the reply!
I just found out that I'm dividing original images into multiple patches when training, and it seems like that there exists patches without boundaries. I guess the code
`if pred_boundary.sum() < 1:`
in forward function states the reason 1 you explained in the previous thread.

Here's the final question for today.. (sorry for making too many issue threads)
If I want to use ABL even when there's no boundary in the image if I change my code to return 0, not None, will it be okay?
I think this might mislead model not giving any penalty when there's no boundary, but just as in the paper we are not using ABL alone, but with other losses, I think I'll work, but not quite sure about it. (Just want to hear your opinion)

Thanks for the so much help! @wangchi95 !!

**Comment by wangchi95** [comment]
> 



> Once again, thanks for the reply! I just found out that I'm dividing original images into multiple patches when training, and it seems like that there exists patches without boundaries. I guess the code `if pred_boundary.sum() < 1:` in forward function states the reason 1 you explained in the previous thread.
> 
> Here's the final question for today.. (sorry for making too many issue threads) If I want to use ABL even when there's no boundary in the image if I change my code to return 0, not None, will it be okay? I think this might mislead model not giving any penalty when there's no boundary, but just as in the paper we are not using ABL alone, but with other losses, I think I'll work, but not quite sure about it. (Just want to hear your opinion)
> 
> Thanks for the so much help! @wangchi95 !!

I’m glad this could be of help to you. Yes, your understanding above is all right!

This is one of the reasons why we don’t use ABL alone, but the main reason is that ABL only supervises the accuracy of boundaries, not specific pixel-level categories. Specifically, for binary segmentation problems, ABL only focuses on the foreground-background segmentation lines, rather than whether each pixel is foreground or background like most loss functions do. This also means that if ABL is used alone, it may reverse the foreground and background.

If you have any further questions, feel free to ask me.

**Comment by Darrenxc** [comment]
> Oh, I got your point. Since the `pred` is logits (we don't know a certain max value), maybe you can concat `-pred` and `pred` to make it work. In addition, you can ignore the ABL in the step with `None` output, which means there detect no obvious prediction boundary. And the boundary is detected here:
> 
> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L65
> 
> And if you are sure there is an obvious boundary there, you can modify this threshold.
> 
> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L38
> 
> 
> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L66
Nice paper and congrat!, I have a question that I am confused about why we need to concat the pred_ and pred together as a logits input of ABL, can you give some more explanations on this, Thank you!

**Comment by wangchi95** [comment]
>> Oh, I got your point. Since the `pred` is logits (we don't know a certain max value), maybe you can concat `-pred` and `pred` to make it work. In addition, you can ignore the ABL in the step with `None` output, which means there detect no obvious prediction boundary. And the boundary is detected here:
>> 
>> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L65
>> 
>> And if you are sure there is an obvious boundary there, you can modify this threshold.
>> 
>> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L38
>> 
>> https://github.com/wangchi95/active-boundary-loss/blob/5820cf377695e678f4bd9c95d0f46235b1a49a98/abl.py#L66
>> 
>> 
> Nice paper and congrat!, I have a question that I am confused about why we need to concat the pred_ and pred together as a logits input of ABL, can you give some more explanations on this, Thank you!

Thank you for your kind words! Concatenating `-pred` and `pred` as logits input to the logits is not the only possible way.

From the foreground probability `p`, we can explicitly get the background probability, which is `1-p`. However, since there is no clear constraint on the scale of the logits, it is difficult to get the background logits only from the foreground logits. While it might be theoretically more reasonable to convert foreground logits to foreground probabilities p using a function such as sigmoid, and then convert 1-p back to logits, this approach is computationally expensive.

The reason why we use logits input instead of probability input is because the joint use of KL and softmax can bring acceleration and avoid [some numerical stability problems.](https://github.com/pytorch/pytorch/issues/32520)

If you don't mind a slight performance loss, you can actually use probabilities as input instead of logits (need to change our kl_div function).

**Comment by RockMan1966** [comment]
> Thank you for your prompt reply，I'll try to modify it as your suggestions.

are you successful?
